### ISSUE TITLE
What is the WCAG definition of "component"?

### ISSUE THREAD

**Issue by jha11y** [w3c/wcag#4252]
The word "component"/"components" is used 92 times in WCAG 2.2, but it does not seemed to be explicitly defined.  

This makes the definition of User Interface Component a little unclear, specific note 3 under this term:

> What is meant by "component" or "user interface component" here is also sometimes called "user interface element". 

Is there a difference between user interface component (interactive control?) vs component (element?)

The word component is used in success criterion that include content that is not interactive such as Success Criterion 1.3.3 Sensory Characteristics:

> Instructions provided for understanding and operating content do not rely solely on sensory characteristics of **components** such as shape, color, size, visual location, orientation, or sound.

**Comment by jha11y** [comment]
related issue: https://github.com/w3c/wcag/issues/4234

**Comment by mbgower** [comment]
First, there is a definition for [user interface component](https://www.w3.org/TR/WCAG22/#dfn-user-interface-components), so in any success criterion or paragraph of normative content where that defined term is used and the word "component" occurs within the same context, "component" is clearly intended to be a synonym. For instance:
> User Interface Components
Visual information required to identify [user interface components](https://www.w3.org/TR/WCAG22/#dfn-user-interface-components) and [states](https://www.w3.org/TR/WCAG22/#dfn-states), except for inactive components or where the appearance of the component is determined by the [user agent](https://www.w3.org/TR/WCAG22/#dfn-user-agents) and not modified by the author;

Second, although there are 92 occurrences of the word "component" in the standard, not all are used in the sense of a user interface component. Some are meant in the abstract non-technology sense. For example:
> - For an overview of how these components of web development and interaction work together, see...
> - Required Components of a Conformance Claim

A better way to assess a potential problem space is where the word "component" appears in **normative** text where it does not appear in association with the defined term "user interface component". There are 5 criteria where this occurs:

1. [2.1.2 No Keyboard Trap](https://www.w3.org/TR/WCAG22/#no-keyboard-trap)
3. [2.4.3 Focus Order](https://www.w3.org/TR/WCAG22/#focus-order)
4. [2.4.13 Focus Appearance](https://www.w3.org/TR/WCAG22/#focus-appearance)
5. [3.2.4 Consistent Identification](https://www.w3.org/TR/WCAG22/#consistent-identification)
6. [1.3.3 Sensory Characteristics](https://www.w3.org/TR/WCAG22/#sensory-characteristics)

Given that No Keyboard Trap, Focus Order, and Focus Appearance are all about interaction, the context seems very clear, and does not seem to create a problem. The language of Consistent Identification states "Components that have the [same functionality](https://www.w3.org/TR/WCAG22/#dfn-same-functionality), which again makes it clear it is about something that can be operated. So the only one that is potentially unclear is the one you flagged, Sensory Characteristics.

As per your quote of its normative text, Sensory Characteristics covers both "understanding and operating". As well, the examples in the Understanding document cover both operation of buttons _and_ a situation where a user is interpretting the information in a table. If "component" was used strictly in the sense of user interface component, it would actually restrict the scope of this requirement. It would only cover instructions about operation. The first example concerning interpreting colour on a non-operable schedule would no longer be valid.

If you think we should add additional context in this Understanding document to clarify it does not apply only to instructions for user interface components, please confirm. Also,  feel free to provide suggestions.

There are also 6 definition where the word occurs without a link to the definition

1. label
2. minimum bounding box
3. name
4. role
5. synchronized media
6. viewport (uses "interface components" without linking to definition)

Almost all of the occurrences of these defined terms happen within the context of user interface components. Label could be considered an exception to this, since it also is used normatively in [Headings and Labels](https://www.w3.org/TR/WCAG22/#headings-and-labels), which are typically not operable. If you believe we should add additional context in this Understanding document, and where, please let us know.

.

**Comment by electronicwoft** [comment]
a further confuision occurs when speaking with developers who commonly use 'component' to mean a named unit of code in the sense of 'widget' or 'module' which may or may not have any bearing on whether someone using the UI perceives such and such as performing a given function. 

A standalone button is a component in the mind of a developer as much as a series of buttons at the bottom of a form because this is how a developer may implement these 'user interface components'. 

The distinction made on the basis of 'functionality' as in SC3.2.4 doesn't resolve the issue of the inconsistent use of 'component' either given that 'functionality' is normatively defined as "processes and outcomes achievable through user action"

It is not the interactivness of functionality, but the possibility that an outcome may be achieved using something other than a particular component or components in the sense of units of code. 

Someone may use the search component on a website to find a particular topic or use the provided navigation, and, in the sense of component as unit of code, they've used two components, but the outcome and therefore the functionality provided by these two units of code is the same. 

This also occurs in the definition of a user interface component: "a part of the content that is perceived by users as a single control for a distinct function".

This 'distinct function' is the way that developers might understand how a unit of code is expected to behave and it is similarly not consistent with the normative definition of functionality.

As the search example illustrates, a user might understand two different units of code as performing the same function whereas it is very common for developers to see function in the sense of 'a single control' like a dropdown as something akin to 'activation behaviour' in HTML.

The equating of 'user interface component' with 'single control' in this definition affirms this connotation.

In my experience, this can lead to lengthy and unnecessary back and forth between evaluators who may know nothing about code who describe accessibility issues experientially and developers who may have little or no understanding of a user interface beyond instantiating a component in a front-end library.

Whatever changes are to be made if any (probably too late for 2.x) should take this equivocality into consideration.

**Comment by jha11y** [comment]
>  Almost all of the occurrences of these defined terms happen within the context of user interface components. Label could be considered an exception to this, since it also is used normatively in [Headings and Labels](https://www.w3.org/TR/WCAG22/#headings-and-labels), which are typically not operable. If you believe we should add additional context in this Understanding document, and where, please let us know.

If in the definition of Label in WCAG (definition included below), the term “component” (used twice) is considered an exception, this seems like there might need to be some guidance in WCAG itself about the term “component”.  This might include possibly a note or a separate definition. Shouldn’t a term used in WCAG either be straight forward with no exceptions or well defined including any exceptions?


> Label definition
> text or other component with a text alternative that is presented to a user to identify a component within web content

There are concerns from the Mobile Accessibility Task Force (MATF) about how the term component fits in with the definition of Label. Particularly when it comes to web links, Do web links have a label? Is a web link considered a “component”? Getting an answer to these questions from the larger Working Group will effect how this task force molds the WCAG success criterion to fit with native mobile content paradigms.

You also mentioned:

> Almost all of the occurrences of these defined terms happen within the context of user interface components.

One of those terms was Name (definition included below)

> Name definition
> text by which software can identify a component within web content to the user

If “component” is synonymous with “use interface component”, does that mean that a Name (Accessible name) is only required for interactive items and not structural or content-based items like landmarks, groupings, tables, etc.?

**Comment by mbgower** [comment]
> Shouldn’t a term used in WCAG either be straight forward with no exceptions or well defined including any exceptions?

Sure, but unfortunately that is not always the reality. The Task Force has to work within the constraints of the existing normative content; it strives to make things closer to what I think we both agree is how things _should_ be. The scrutiny thus far has turned up two situations where I think some context is needed (Sensory Characteristics and Heading and Labels). I think we can provide clarity there. How _other_ people use "component" is obviously outside TF control.


The occurrences of the defined term _name_ seem to be restricted to UIC situations. If you find one that is not, please surface.  It is important to note that a word like "name" may be used normatively (and it should always reference the definition in its first use in a normative context) and non-normatively. For instance, the word "name" appears in lots of situations where it is not specifically used to mean the defined term [name](https://www.w3.org/TR/WCAG22/#dfn-name).

> If “component” is synonymous with “use interface component”, does that mean that a Name (Accessible name) is only required for interactive items and not structural or content-based items like landmarks, groupings, tables, etc.?

I think a better way to put that is that defined uses of _name_ lead to 4.1.2, which was added to knit ARIA into the WCAG fabric for UICs beyond the HTML elements available. So from that perspective, I'd answer "yes".

But it's important to emphasize that 1.3.1 Info and Relationships also deals with "Information, [structure](https://www.w3.org/TR/WCAG22/#dfn-structure), and [relationships](https://www.w3.org/TR/WCAG22/#dfn-relationships) conveyed through [presentation](https://www.w3.org/TR/WCAG22/#dfn-presentation)". You'll note that "name" doesn't appear in 1.3.1 normative text, although different technologies might use attributes called "name" to achieve 1.3.1. Before ARIA came along, there were still programmatic ways in which you could designate table structure and some kinds of groupings. Landmarks are typically also addressed by 1.3.1, because a region is presentational, not operational (although it can obviously contain UICs). 

I'm opening issues to deal with the 2 items I identified so far that require action. If you think others specifically need addressing, please indicate.

**Comment by innoticFR** [comment]
I had a related question in the discussion thread here: https://github.com/w3c/wcag/discussions/5032, which I have now closed in favour of this issue.

I understand that the term "component", when used in isolation, is not necessarily tied to the defined term "user interface component" in WCAG.

However, what remains unclear to me is how consistently the term "user interface component" is intended to apply across different success criteria, particularly in cases where the usage seems conceptually different.

For example, do SC 2.4.10 and SC 4.1.2 refer to the same concept of "user interface component"?

> 2.4.10: This success criterion covers sections within writing, not [user interface components](https://www.w3.org/TR/WCAG22/#dfn-user-interface-components)

> 4.1.2: For all [user interface components](https://www.w3.org/TR/WCAG22/#dfn-user-interface-components) (including but not limited to form elements, links, and components generated by scripts), the [name](https://www.w3.org/TR/WCAG22/#dfn-name) and [role](https://www.w3.org/TR/WCAG22/#dfn-role) can be programmatically determined.

My question is whether SC 2.4.10 and SC 4.1.2 are using "user interface components" in the same defined sense, or whether the term is being applied differently between these success criteria.

If it is intended to be the same definition in both cases, then I am struggling to understand how SC 2.4.10 should be interpreted in relation to individual controls, since it appears to refer to structural sections of content rather than operable interface elements.

If it is not intended to be the same sense, could this distinction be made clearer in the wording or in the Understanding documentation?

For example, is a "user interface component", according to this definition, always intended to refer to an individual interactive control, and not a broader structural element of content?

As an illustration of this question: does the definition exclude the idea that a user interface component could itself contain other user interface components (i.e., nesting of UICs), or is that interpretation also valid under WCAG terminology?

Thank you for any clarification.

**Comment by patrickhlauke** [comment]
@innoticFR at a high level, i'd preface this by saying that i *suspect* the terminology was a product of the time when this was originally written (almost 20 years ago?) ... and the idea back then that there's usually a clear distinction between the writing/content of a page, and discrete interactive controls like buttons, links, form fields, etc. that "clear" distinction has become more and more blurry over the years - culminating in cases where arguably an entire page/site can be seen as one big interactive web app where almost anything on screen can be argued to be an interactive UIC to some extent...

I think to help guide the discussion, I'd straight out ask: what specifically are you finding hard to categorise? what scenario do you have in mind that makes you unsure about which SC can be applied, in what situation?

from my own experience, I know a lot of auditors use 4.1.2 Name, Role, Value as the one SC where they hang any "this thing lacks a `role`" problems on. however, for some (non-interactive) structures, it's arguably more a problem that would fall under 1.3.1 Info and Relationships (assuming the visual presentation of something hints at the fact it has some specific underlying function or structure, and that this needs to be exposed to non-visual users, otherwise they'd miss out on meaning/context)

**Comment by innoticFR** [comment]
Thanks for this quick answer, it makes sense. "User interface components" 20 years ago were mainly understood as interactive elements, with less emphasis on semantics across all content.

Looking at [what was considered a "control" in HTML4](https://www.w3.org/TR/html401/interact/forms.html) helps clarify this.

However, the current wording in WCAG seems to introduce inconsistencies:

> 1.1.1: If non-text content is a control or accepts user input  

→ suggests that an input is not a control

> 1.3.5: Input Purposes for User Interface Components  

→ implies that an input is a user interface component

> Definition: user interface component = "a part of the content that is perceived by users as a single control for a distinct function"  

→ suggests that a user interface component is a control

This leads to a tension between:
- "control" as used in SC 1.1.1 (possibly narrower, closer to its general dictionary meaning as an action-triggering element)
- "control" as used in the definition (broader, closer to HTML form controls)

If "user interface component" is understood as an HTML "control", this interpretation seems consistent across most success criteria, except possibly SC 2.4.10.

Finally, note 3 adds further confusion:

> "What is meant by 'component' or 'user interface component' here is also sometimes called 'user interface element'."

In practice, "user interface element" is a broader term, and not necessarily equivalent to "user interface component".

The reason I am raising this is that, when expressing these concepts in another language, the underlying usages can differ significantly. For example, the equivalent of "component" often carries a broader meaning, while the equivalent of "control" tends to be understood more narrowly  (e.g. controls used to trigger actions). As a result, a direct, word-for-word translation of the definitions can unintentionally shift their meaning, rather than preserve the intended scope.

**Comment by electronicwoft** [comment]
'distinct function'  is also a problematic  phrase in this context and least of all because of the prepending of 'distinct'.

If a user interface component is understood as a control, then its 'function' describes  the activation behaviour of that control rather than as an 'outcome achievable through user action' as functionality is defined in the glossary.

That is, select, accept text, click, etc. rather than whatever process or outcome activation may trigger. 

As I said above, the differences in the way that people using an interface and the people developing the interface talk about these kinds of things ... for the developer, it's a dropdown,. for a user, it's a way of choosing their favourite icecream.

That auditors commonly  see 4.1.2 as the clearest example of component in the sense of control is possibly evidence of the general ambiguity of component in WCAG and a tendency towards certainty by understanding component within the strictures of interactive HTML elements.

And, conversely, why most don't comprehend that which falls under 1.3.1 as non-interactive components despite being perceived by users as having a 'distinct' function'.

None of the normative terms in WCAG 2.x operate in a vacuum and humans don't consume them as if they did in general so it makes little sense in focusing on 'component' and 'user interface component' without also taking into account the meaning of the terms around them like function and functionality.
